engineapi/backup_monitor.go

```diff
@@ -72,7 +72,7 @@ func NewBackupMonitor(logger logrus.FieldLogger, ds *datastore.DataStore, backup
 	}
 
 	// Call engine API snapshot backup
-	if backup.Status.State == longhorn.BackupStateNew {
+	if backup.Status.State == longhorn.BackupStateNew || backup.Status.State == longhorn.BackupStatePreparing {
```

Comment:
Hi @c3y1huang 
Yes, it will be called only once,

In the previous version, since we didn't add `preparing` for waiting snapshot/engine to be ready
so the state will be `""`, we then use this condition to trigger backup grpc call

Now, since we want it to wait for snapshot/engine to be ready, we use `"preparing"` to hold it
After the snapshot/engine is ready, the state will still be `"preparing"` in this condition. 
Thus we need to allow `Preparing` to trigger backup

That is, for state transition
```
if snapshot/engine is already ready when backup CR is created
=> "" => "in-progress" => "completed"

if snapshot/engine is not ready
=> "" => "preparing" => "in-progress" => "completed" 
```

Reply:
After the backup is triggered, the state then be override by the monitor

Reply:
I see. Thank you for the explanation.

Is there any concern about using the Pending state? You mentioned that `Pending state has been used for another purpose` in the description, but I feel we might not need a new state for this.

1. A Backup CR is created.
1. The controller checks the snapshot/engine state before creating the monitor. 
1. If the snapshot/engine is not ready, an error is returned, and the controller reconciles again.
1. If the snapshot/engine is ready, the controller creates the backup monitor and that calls engineClientProxy.SnapshotBackup.

I feel like we could update the state to Pending instead in step 3 and then:
```
// Call engine API snapshot backup
	if backup.Status.State == longhorn.BackupStateNew || backup.Status.State == longhorn.BackupStatePending {
```

Reply:
Hi @c3y1huang 
No we can't.
Because pending state means it already make the grpc call but gets `DeadlineExceeded`
It mark the state as `Pending`  and don't return the error, instead, it keeps creating the monitor and see if the state changes 

```
func NewBackupMonitor(......) (*BackupMonitor, error) {
	// Call engine API snapshot backup
	if backup.Status.State == longhorn.BackupStateNew {
		// volumeRecurringJobInfo could be "".
		volumeRecurringJobInfo, err := m.getVolumeRecurringJobInfos(ds, volume)
		if err != nil {
			return nil, err
		}
		// put volume recurring jobs/groups information into backup labels and it would be stored in the file `volume.cfg`
		if volumeRecurringJobInfo != "" {
			backup.Spec.Labels[types.VolumeRecurringJobInfoLabel] = volumeRecurringJobInfo
		}
		_, replicaAddress, err := engineClientProxy.SnapshotBackup(engine, backup.Spec.SnapshotName, backup.Name,
			backupTargetClient.URL, volume.Spec.BackingImage, biChecksum, string(compressionMethod), concurrentLimit, storageClassName,
			backup.Spec.Labels, backupTargetClient.Credential)
		if err != nil {
			if !strings.Contains(err.Error(), "DeadlineExceeded") {
				m.logger.WithError(err).Warn("Cannot take snapshot backup")
				m.Close()
				return nil, err
			}

			// [Workaround]
			// Special handling the RPC call return code DeadlineExceeded, mark it as Pending state.
			// The snapshot backup initialization _probably_ succeeded in the replica sync agent server.
			// Use the backup monitor routine to monitor the backup status stays in Error state or change
			// to InProgress/Completed state with the maximum retry count mechanism.
			// [TODO]
			// Since API engineclient.SnapshotBackup is not idempotent, this controller cannot blindly
			// retry the call when error DeadlineExceeded is triggered.
			// Instead, it has to mark the backup as a kind of special state Pending, then relies on the
			// backup monitor routine periodically checking if the backup creation actually started.
			// After making the API call idempotent and being able to deprecate the old version
			// engine image, we can remove this part.
			// https://github.com/longhorn/longhorn/issues/3545
			m.logger.WithError(err).Warnf("Snapshot backup timeout")
			backup.Status.State = longhorn.BackupStatePending
		}

		m.backupStatus = backup.Status
		m.replicaAddress = replicaAddress
	}

	// Create a goroutine to monitor the replica backup state/progress
	go m.monitorBackups()

	return m, nil
}
```

If we change the condition to 
`if backup.Status.State == longhorn.BackupStateNew || backup.Status.State == longhorn.BackupStatePending {}`

Then the backup grpc call will be called multiple times because the state might keep in `Pending` and it then keeps creating multiple goroutines for monitoring every time.

The state change starts when the backup grpc is called and is updated by the monitor
and we need to make sure it is called only once.

So we need a new state to present the status of **"before the backup is called"**

Reply:
After discussed with @c3y1huang 
Yes, we can reuse the `Pending` state
There are two cases for `Pending` state

1. Backup is waiting for Engine and Snapshot before calling the grpc
     - in the next reconcile, if everything is ready, it will call the grpc and enable the monitor. The state transition will be handled by monitor
2. Backup grpc is called but `DeadlineExceeded`. In this case, the monitor is enabled. Thus the state transition will also be handled by monitor. Since the monitor is created, so it won' call the grpc again because we check if monitor exists first.